### DISCUSSION TITLE
sleap processes groups of videos

### DISCUSSION THREAD

**Discussion by chrisch12138** [talmolab/sleap#1297] [Help!] [answered]
Hi,
I recently needed to use Sleap to process my behavioral data. My data is 14 five-hour videos that recorded mice group social behaviors. Essentially we would like to label a few frames from each video, and train on frames across all 14 videos and predict all of the videos. The problem is each of our video is about 70 GB, and it would slow down the processing speed of Sleap if we uploaded all videos in one sleap project and labeled/predicted too many frames. We are wondering if anyone has any suggestions on how to efficiently label, train and predict all videos. Thanks!

**Comment by chrisch12138** [comment]
Is there a way to save user-labeled frames so that we can select and add a few frames from each video and train on all of them together?

**Comment by roomrys** [comment]
Hi @chrisch12138,

### TLDR
Have you tried generating a subset of frames to predict/label on via the Labeling Suggestions?

---

As you have noticed, the GUI slows down proportional to the number of *labeled* frames (this includes both predicted and user-labeled frames). Thus, while iterating over the human-in-the-loop training, we recommend generating a subset of frames to predict/label on (from each video) via the Labeling Suggestions dock. It's important to sample across videos for the Labeling Suggestions in order for the model to pick up on differences across pixel intensity/background.

Once you have a model that performs well, I would suggest creating a project for each video seeing as you have ample frames in each video. Then, I would use the trained model to perform one final inference/prediction on each video/project. From there, you can start proofreading which will be much faster to correct poses if the projects are split. Related: #757.

Let us know if the Labeling Suggestions method helps!

Thanks,
Liezl

**Reply by chrisch12138** [reply]
Hi Liezl,

Thanks for your reply! Yes I have been trying to generate a few frames from each video via Labeling Suggestions. Currently I don't have a lot of user-labeled frames, just about 100, and I don't have any predicted frames. What I noticed is that the GUI currently lags a lot compared to the past when I separated each video as a single project.
I was wondering whether the lagging is coming from the large number of videos and size of the video we have in a single project. If it is, then is there a way to only save user-labeled frames from one project and just import that for another individual project (video) as we try to train the model across all videos? However you mentioned the GUI slows down only when there are a lot of predicted/labeled frames, which is not the case here... 

Thanks for your help! 

Chris

**Reply by roomrys** [reply, marked as answer]
Hi @chrisch12138,

Hmm, I expected it to be a problem with predicting on the entire video during the human-in-the training loop, but seeing as that isn't the case...

Which extensions do your videos have (mp4, npy, etc)?

SLEAP creates it's own cache to store frames for certain extensions, but for mp4/avi extensions we use `cv2.VideoCapture` to open a reader for each video. 

### Possible Cause A: Memory Overload from `VideoCapture` Buffer
This reader remains open while the project is open and has a default buffer size of 38 frames. So, for each video you have in the project, their will be 38 frames in memory which might be what is causing the slowdown (especially if you have high resolution videos). This would also explain the slowdown being proportional with number of videos. 

### Possible Cause B: Jump-Seeking Leads to More Frequent Buffer Recreation
Another cause might be the jump-seeking seen in SLEAP. Displaying non-sequential frames makes it more likely that the frame we want is not currently in the buffer. Thus, the buffer is not actually used which causes a lag. But, this would not explain the slowdown being proportional to the number of videos.

### Action Items
I'll need do some of my own tests to verify the above, but a possible solution we could implement would be to reduce the buffer size (seeing as it is not very useful for jump-seeking). Until this is solved, you could trim your videos using ffmpeg and use the shorter clips in the project to sample training frames from:
```
ffmpeg -i input.mp4 -ss 00:05:20 -t 00:10:00 -c:v copy -c:a copy output1.mp4
```
> Reminder to save to a different filepath/name for the output - otherwise you will overwrite the existing video. 
Also, the `sleap` conda environment already has `ffmpeg` installed, so you can just activate the `sleap` environment to run the above command.

Thanks for bringing this to our attention!
Liezl

**Reply by chrisch12138** [reply]
Hi @roomrys ,

Yes our videos are in .avi formats. And thanks, I'll try to trim and use the shorter clips for now.

Thanks!

Chris
